File: ts/binary_tree_display.ts
class BinaryTreeNode {
    key: any;
    left: BinaryTreeNode | null;
    right: BinaryTreeNode | null;
    id: string | null;

    constructor(key: any, left: BinaryTreeNode | null = null, right: BinaryTreeNode | null = null) {
        this.key = key;
        this.left = left;
        this.right = right;
        this.id = null;
    }

    to_dot(dot_filepath: string, rankdir: string): void {
        let nodes_lines = `\t"${this.key}" [label=${this.key}, fontname="Roboto", shape="rectangle", style="rounded", root=true]\n`;
        let edges_lines = '';

        let to_visit: BinaryTreeNode[] = [this];
        this.id = this.key;

        let usedids: string[] = [this.id];
        while (to_visit.length > 0) {
            let node = to_visit.pop();
            let left = node.left;
            let right = node.right;
            if (left) {
                let leftid = left.key;
                while (usedids.includes(leftid)) {
                    leftid += '_';
                }
                left.id = leftid;
                usedids.push(left.id);

                nodes_lines += `\t"${left.id}" [label="${left.key}", fontname="Roboto", shape="rectangle", style="rounded"]\n`;
                edges_lines += `\t"${node.id}" -> "${left.id}" [dir="back"]\n`;

                to_visit.unshift(left);
            }
            if (right) {
                let rightid = right.key;
                while (usedids.includes(rightid)) {
                    rightid += '_';
                }
                right.id = rightid;
                usedids.push(right.id);

                nodes_lines += `\t"${right.id}" [label="${right.key}", fontname="Roboto", shape="rectangle", style="rounded"]\n`;
                edges_lines += `\t"${node.id}" -> "${right.id}" [dir="back"]\n`;

                to_visit.unshift(right);
            }
        }

        const fs = require('fs');
        fs.writeFileSync(dot_filepath, 'digraph tree {\n');
        fs.appendFileSync(dot_filepath, `\trankdir=${rankdir}\n`);
        fs.appendFileSync(dot_filepath, nodes_lines);
        fs.appendFileSync(dot_filepath, '\n');
        fs.appendFileSync(dot_filepath, edges_lines);
        fs.appendFileSync(dot_filepath, '}');
    }
}

File: binary_tree_display.py

class BinaryTreeNode:
    def __init__(self, key, left=None, right=None):
        self.key = key
        self.left = left
        self.right = right
        # ID is for graphviz
        self._id = None

    def to_dot(self, dot_filepath, rankdir="BT"):
        # The 'lines' to write to the file - one section for nodes and one for edges
        nodes_lines = f'\t"{self.key}" [label={self.key}, fontname="Roboto", shape="rectangle", style="rounded", root=true]\n'
        edges_lines = ''
        # Simple breadth-first search - there basically is no difference 
        # between depth-first and breadth-first, I think the latter is just 
        # easier to understand.
        to_visit = [self]
        self._id = self.key
        # We don't want any ID repeats
        used_ids = [self._id]
        while to_visit:
            node = to_visit.pop()
            left = node.left
            right = node.right
            if left:
                # Adding underscores to the left node's key until it is unique from any other
                left_id = left.key
                while left_id in used_ids:
                    left_id += '_'
                
                left._id = left_id
                used_ids.append(left._id)
                # Add to nodes_lines and edges_lines to write to file
                nodes_lines += f'\t"{left._id}" [label="{left.key}", fontname="Roboto", shape="rectangle", style="rounded"]\n'
                edges_lines += f'\t"{node._id}" -> "{left._id}" [dir="back"]\n'
                # Visit this node later
                to_visit.insert(0, left)
                
            if right:
                # Same here, add underscores until unique
                right_id = right.key
                while right_id in used_ids:
                    right_id += '_'
                
                right._id = right_id
                used_ids.append(right._id)
                # Add to write
                nodes_lines += f'\t"{right._id}" [label="{right.key}", fontname="Roboto", shape="rectangle", style="rounded"]\n'
                edges_lines += f'\t"{node._id}" -> "{right._id}" [dir="back"]\n'
                # Visit later
                to_visit.insert(0, right)
        
        # Write lines to file
        with open(dot_filepath, 'w') as writefile:
            writefile.write('digraph tree {\n')
            writefile.write(f'\trankdir={rankdir}\n')
            writefile.write(nodes_lines)
            writefile.write('\n')
            writefile.write(edges_lines)
            writefile.write('}')
